blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/credential/BlueOceanCredentialsProvider.java

```diff
@@ -0,0 +1,261 @@
+import hudson.model.User;
+import hudson.security.Permission;
+import hudson.util.ListBoxModel;
+import io.jenkins.blueocean.rest.impl.pipeline.Messages;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.acegisecurity.Authentication;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link CredentialsProvider} to serve credentials stored in user store.
+ *
+ * It works by looking for {@link FolderPropertyImpl} on a folder. This should allow
+ * using such user scoped properties with {@link jenkins.branch.OrganizationFolder} or {@link jenkins.branch.MultiBranchProject}
+ * or even a Folder.
+ *
+ * @author Stephen Connoly
+ * @author Vivek Pandey
+ */
+@Extension(ordinal = 99999)
+public class BlueOceanCredentialsProvider extends CredentialsProvider {
+    @Nonnull
+    @Override
+    public <C extends Credentials> List<C> getCredentials(@Nonnull Class<C> type, @Nullable ItemGroup itemGroup,
+                                                          @Nullable Authentication authentication) {
+        if (itemGroup instanceof AbstractFolder) {
+            return getCredentials(type, (AbstractFolder)itemGroup);
+        }
+        return Collections.emptyList();
+    }
+
+    private static <C extends Credentials> List<C> getCredentials(Class<C> type, AbstractFolder folder){
+        FolderPropertyImpl prop = (FolderPropertyImpl)folder.getProperties().get(FolderPropertyImpl.class);
+        if (prop != null) {
+            User user = User.get(prop.getUser(), false, Collections.emptyMap());
+            if(user != null){
+                Authentication auth = user.impersonate();
+                return CredentialsMatchers.filter(
+                        CredentialsProvider.lookupCredentials(
+                                type,
+                                Jenkins.getInstance(),
+                                auth,
+                                Collections.<DomainRequirement>emptyList()
+                        ),
+                        CredentialsMatchers.allOf(CredentialsMatchers.withId(prop.getId()),
+                                CredentialsMatchers.withScope(CredentialsScope.USER))
+                );
+            }
+        }
+        return Collections.emptyList();
+    }
+
+    @Nonnull
+    @Override
+    public <C extends IdCredentials> ListBoxModel getCredentialIds(@Nonnull Class<C> type,
+                                                                   @Nullable ItemGroup itemGroup,
+                                                                   @Nullable Authentication authentication,
+                                                                   @Nonnull List<DomainRequirement> domainRequirements,
```

Comment:
Given that you are creating the credentials in a domain in the user store, we should probably be applying the same domain validation in the `getCredentials` and `getCredentialIds` methods. This would mean storing either the `URL` or the `DomainSpecification` as a field in the `FolderPropertyImpl`